### PR TITLE
Object lock and unlock

### PR DESCRIPTION
**Pull request by didoda**
This provides lock and unlock on objects.

Rif: https://chialab.atlassian.net/browse/BEM-24

### CODE REVIEW

File: src/Controller/LockController.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    protected function lock(bool $locked): void
+    {
+        $type = $this->request->getParam('object_type');
+        $id = $this->request->getParam('id');
+        $meta = compact('locked');
+        $data = compact('id', 'type', 'meta');
+        $payload = json_encode(compact('data'));
+        try {
+            $this->apiClient = ApiClientProvider::getApiClient();
```
Comment (stefanorosanelli) [inline src/Controller/LockController.php]: Not necessary - `$this->apiClient` already instantiated by `AppController`

File: src/Controller/LockController.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
+<?php
+namespace App\Controller;
```
Comment (stefanorosanelli) [inline src/Controller/LockController.php:13]: missing copyright header

File: src/Template/Pages/Modules/view.twig [hunk 3]
```diff
@@ -25,6 +25,7 @@
             })|raw }}
 
             {{ Form.hidden('id', {'value': (object) ? object.id : resource.id})|raw }}
+            {{ Form.hidden('locked', {'value': (object) ? object.meta.locked : false})|raw }}
```
Comment (stefanorosanelli) [inline src/Template/Pages/Modules/view.twig]: are we using this `locked` form field anywhere? can we remove it?

File: src/Template/Pages/Modules/view.twig [hunk 4]
```diff
@@ -115,8 +116,26 @@
...
 
+            {# Append "Lock/Unlock" #}
+            {% if object.id and Perms.canLock() %}
+                {% if object.meta.locked %}
+                    {% do _view.append('module-buttons', Html.link(
+                        __('Unlock'),
+                        {'_name': 'lock:remove', 'object_type': objectType, 'id': object.id},
+                        {'class': 'button button-outlined button-outlined-hover-module-' ~ currentModule.name }
+                    )) %}
+                    {{ Form.hidden('locked', {'value': object.meta.locked})|raw }}
```
Comment (stefanorosanelli) [inline src/Template/Pages/Modules/view.twig]: are we using this `locked` form field anywhere?

File: src/Controller/LockController.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+{
+    /**
+     * Add lock
+     *
+     * @return void
+     */
+    public function add(): void
+    {
+        $this->lock(true);
+        $this->redirect($this->referer());
```
Comment (stefanorosanelli) [inline src/Controller/LockController.php]: the proper way to redirect is to use 
```php 
  return $this->redirect(..);
``` 
same below

File: src/Template/Element/Modules/index_table_row.twig [hunk 6]
```diff
@@ -89,6 +91,8 @@
                         'class': buttonClasses,
                     }
                 )|raw}}
+            {% elseif object.meta.locked %}
+                [ {{ __('Locked') }} ]
```
Comment (stefanorosanelli) [inline src/Template/Element/Modules/index_table_row.twig]: I'd leave it empty for now - or use a `locked` icon

File: src/Controller/LockController.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        $payload = json_encode(compact('data'));
+        try {
+            $this->apiClient = ApiClientProvider::getApiClient();
+            $this->apiClient->patch(
+                sprintf('/%s/%s', $type, $id),
+                $payload,
+                ['Content-Type' => 'application/json']
+            );
+        } catch (BEditaClientException $ex) {
+            $this->loadComponent('App.Flash', ['clear' => true]);
```
Comment (stefanorosanelli) [inline src/Controller/LockController.php]: maybe an additional error log?

`$this->loadComponent()` not necessary, already provided by `AppController`

File: src/Controller/AppController.php [hunk 8]
```diff
@@ -244,6 +244,14 @@ function ($item) {
                 return ['name' => $category];
             }, $data['categories']);
         }
+
+        // locked
+        $locked = (string)Hash::get($data, 'locked');
```
Comment (stefanorosanelli) [inline src/Controller/AppController.php]: Do we need this? Is `locked` form field necessary?
(see also comments below)

**Review by stefanorosanelli — CHANGES_REQUESTED** [review]

File: src/Template/Pages/Modules/view.twig (hunk 4 above)
Comment (didoda) [reply src/Template/Pages/Modules/view.twig]: yes, it's used to remove "status" and "uname" from payload, on data save, when object is locked

File: src/Controller/AppController.php (hunk 8 above)
Comment (didoda) [reply src/Controller/AppController.php]: this provides the server side logic, to remove from payload "status" and "uname", when saving a locked object. not necessary, but useful IMHO

**Review by stefanorosanelli — APPROVED** [review]

**Pull request merged**